feat!: validate Avro schemas included in AsyncAPI docs

dalelane opened the pull request:
This commit adds validation to the avro-schema-parser so that
invalid Avro schemas will be rejected.

I had to fix a few invalid Avro schemas in the tests (which were
using the type 'uuid' which isn't in the Avro spec) so this should
potentially be considered a major breaking change to the parser.

Closes: #37

Signed-off-by: Dale Lane <dale.lane@uk.ibm.com>

JQrdan reviewed (dismissed):
This looks good! Thanks @dalelane

magicmatatjahu commented on tests/schemas/Person-broken.avsc:
```diff
@@ -0,0 +1,21 @@
+  "type": "record",
+  "fields": [
+    {"name": "name", "type": "string", "example": "Donkey"},
+    {"name": "age", "type": ["null", "int"], "default": null, "example": "123"},
```
I have one question. If types for that field are `["null", "int"]`, so example should be written as number, not as string, am I right - at least in JSON Schema we should write it in this way, is it different with avro?

dalelane replied:
That line came from the existing example at https://github.com/asyncapi/avro-schema-parser/blob/61cbd19f972b1a7cbc8daf3f137c40e1d26aeaa5/tests/schemas/Person-1.9.0.avsc#L6 

but excuses-aside, yes - I think you're right - that should be an int, like `"example": 123` 

This needs fixing in:
- https://github.com/asyncapi/avro-schema-parser/blob/master/tests/schemas/Person-1.9.0-additional-attributes.avsc#L9 
- https://github.com/asyncapi/avro-schema-parser/blob/master/tests/schemas/Person-1.9.0-namespace.avsc#L7
- https://github.com/asyncapi/avro-schema-parser/blob/master/tests/schemas/Person-1.9.0.avsc#L6

magicmatatjahu replied:
I wonder if someone made a mistake then and it needs to be corrected in all places. It would be best to ask someone who is an expert in avro. As far as I'm concerned this is the only thing that caught my eye, the rest is ok :)

M3lkior replied:
this is right but because the `example` attribute is not an official AVRO attribute, this is not critical and i guess that the validation should pass if the example value is not correctly typed (because it is an additional attribute ...)

dalelane replied:
created https://github.com/asyncapi/avro-schema-parser/issues/97 for this

M3lkior commented on to-json-schema.js:
```diff
@@ -145,6 +155,8 @@ async function convertAvroToJsonSchema(avroDefinition, isTopLevel) {
     return jsonSchema;
   }
 
+  validateAvroSchema(avroDefinition);
```
the `validateAvroSchema` can be placed before the `if (isUnion)` test ? (before ligne 143)

dalelane replied:
yeah, I think that would work too

M3lkior replied:
this can avoid executing the previous instructions :)

magicmatatjahu commented:
For me is ok, but I'm not a Codeowner so you need to wait for @derberg @fmvilas or @M3lkior

M3lkior commented:
> For me is ok, but I'm not a Codeowner so you need to wait for @derberg @fmvilas or @M3lkior

Sorry, i will work on this tomorrow :)

derberg commented:
@dalelane I see a remark about breaking change. Basically, we are afraid that enabling validation can cause failing parsing for others?

the Avro parser is already used by many, and in production, and is in a good shape. We can release 1.0 imho.

So even if we are not 100% sure if it is a breaking feature, we can play safe

thoughts?

dalelane commented:
> @dalelane I see a remark about breaking change. Basically, we are afraid that enabling validation can cause failing parsing for others?
> 
> the Avro parser is already used by many, and in production, and is in a good shape. We can release 1.0 imho.
> 
> So even if we are not 100% sure if it is a breaking feature, we can play safe
> 
> thoughts?

yeah, I think that would be good

To be fair, I think the risk is very small - because if someone really did have an invalid Avro schema in this way, most other Avro tools/clients/libraries I'm aware of would've already rejected it. 

But, there is a small chance that someone has a slightly-invalid Avro schema in an AsyncAPI doc that today would parse correctly - and upgrading the parser would cause it to no longer parse. The Avro schemas we had in the tests are examples of how that can happen, as I needed to fix them! :-)

derberg commented:
Normally we do not treat these things as breaking changes. For example, the main JS Parser doesn't have 100% validation coverage, and when we add new ones, like tags validation and at some point of time examples validation, we do not make a major release.

But yeah, here we are in this good position that we can actually do a major if we all think it is ok. Count on my vote

M3lkior reviewed (commented):
@dalelane it could be nice to add a new section in the `Features` chapter in the `README` file to indicate that we now provide avro validation during the processing ? 
Furthermore, if we decide to release as a breaking change version, don't forget to reword your rebased commit with the `!` char, according to the [conventionalcommits.org](https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with--to-draw-attention-to-breaking-change) specification :)

derberg commented:
yes, to release major, title must be `feat!: validate Avro schemas included in AsyncAPI docs` and make sure you merge&squash with the same commit message

dalelane commented:
added an explanation to README, and squashed-and-rebased
hopefully we're good to go now

M3lkior commented:
> added an explanation to README, and squashed-and-rebased hopefully we're good to go now

Yep it seems ok for me :)

dalelane commented:
@derberg brill, thanks - I've squashed and rebased, updated the commit title, and made the README change - think I just need a re-review please then we're good to go

derberg commented:
@dalelane oh, sorry for misunderstanding, I just meant that when you will click button `squash and merge` you need to take care of these things, not that you need to do it on local 😅 

anyway, tests hang there in the system and won't run. Could you push an empty commit maybe to trigger checks please

derberg commented:
Congrats! please only remember folks that even though you are committers you should work on forks like every other contributor, not create branches in the upstream. I recommend [this instruction](https://github.com/asyncapi/.github/blob/master/git-workflow.md) to set up your fork to work with upstream seamlessly

dalelane commented:
ah, I didn't realise - thanks for the heads-up, will do that next time
